libraries/concurrent-merkle-tree/src/cmt.rs

```diff
@@ -0,0 +1,447 @@
+use crate::{
+    error::CMTError,
+    state::{ChangeLog, Node, Path, EMPTY},
+    utils::{empty_node, empty_node_cached, fill_in_proof, hash_to_parent, recompute},
+};
+use bytemuck::{Pod, Zeroable};
+pub(crate) use log_compute;
+pub(crate) use solana_logging;
+
+#[cfg(feature = "sol-log")]
+use solana_program::{log::sol_log_compute_units, msg};
+
+/// Enforce constraints on max depth and buffer size
+#[inline(always)]
+fn check_bounds(max_depth: usize, max_buffer_size: usize) {
+    assert!(max_depth < 31);
+    // This will return true if MAX_BUFFER_SIZE is a power of 2 or if it is 0
+    assert!(max_buffer_size & (max_buffer_size - 1) == 0);
+}
+
+/// Conurrent Merkle Tree is a Merkle Tree that allows
+/// multiple tree operations targeted for the same tree root to succeed.
+///
+/// In a normal merkle tree, only the first tree operation will succeed because the
+/// following operations will have proofs for the unmodified tree state. ConcurrentMerkleTree avoids
+/// this by storing a buffer of modified nodes (`change_logs`) which allows it to implement fast-forwarding
+/// of concurrent merkle tree operations. 
+///
+/// As long as the concurrent merkle tree operations
+/// have proofs that are valid for a previous state of the tree that can be found in the stored
+/// buffer, that tree operation's proof can be fast-forwarded and the tree operation can be
+/// applied. 
+///
+/// There are two primitive operations for Concurrent Merkle Trees:
+/// [set_leaf](ConcurrentMerkleTree:set_leaf) and [append](ConcurrentMerkleTree::append). Setting a leaf value requires
+/// passing a proof to perform that tree operation, but appending does not require a proof.
+///
+/// An additional key property of ConcurrentMerkleTree is support for [append](ConcurrentMerkleTree::append) operations, 
+/// which do not require any proofs to be passed. This is accomplished by keeping track of the
+/// proof to the rightmost leaf in the tree (`rightmost_proof`).
+#[derive(Copy, Clone)]
+pub struct ConcurrentMerkleTree<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> {
+    pub sequence_number: u64,
+    /// Index of most recent root & changes
+    pub active_index: u64,
+    /// Number of active changes we are tracking
+    pub buffer_size: u64,
+    /// Proof for respective root
+    pub change_logs: [ChangeLog<MAX_DEPTH>; MAX_BUFFER_SIZE],
+    pub rightmost_proof: Path<MAX_DEPTH>,
+}
+
+unsafe impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> Zeroable
+    for ConcurrentMerkleTree<MAX_DEPTH, MAX_BUFFER_SIZE>
+{
+}
+unsafe impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> Pod
+    for ConcurrentMerkleTree<MAX_DEPTH, MAX_BUFFER_SIZE>
+{
+}
```

Comment:
Considering all of the inner types should be safe to implement `Pod` / `Zeroable`, you should be good to go with this one too and avoid the `unsafe`!

Reply:
Mentioned in the previous comment, but Pod / Zeroable unfortunately doesn't work for types with const generic parameters. It probably can if the bytemuck library is patched to do so. There's a lot of tricky edge cases to consider with architecture specific alignment though. For example:
```
struct Foo {
    a: u64,
    b: u128,
    c: u64,
}
```
is size 32 on most x86 and 48 on aarch64, because `b` needs to always fall on an address that is a multiple of 16.

This makes things like
```
struct Bar<T, const N: usize> {
    a: [T, N]
}
```
quite difficult to generalize, but it's certainly possible to have the library execute different codepaths based on the target_arch

Reply:
Yeah, looks like you'd have to declare them as packed to make it work: https://github.com/Lokathor/bytemuck/issues/70 -- might even be safer to do that!

Reply:
Not sure how well this meshes with the rBPF virtual machine. Certainly not an expert, but historically I have not had a good time interacting with objects that are not byte aligned. I think this unsafe impl is completely safe on 64-bit systems like the Solana VM given that everything is padded correctly. There’re weird issues that arise when mutating a packed but unaligned object in a view that this completely side steps

Reply:
Ah gotcha, no worries then